File: crates/aether-evals/src/agents/fake_agent.rs
use super::agent::{Agent, AgentRunResult, RunError};
use crate::Task;
use aether_core::events::{AgentEvent, ToolEvent, TurnOutcome};
use async_stream::try_stream;
use futures::Stream;
use llm::ToolCallResult;
use std::path::{Path, PathBuf};
use tokio::fs::{create_dir_all, write};

#[derive(Clone)]
pub struct FakeAgent {
    events: Vec<AgentEvent>,
    file_writes: Vec<(PathBuf, String)>,
    workspace: Option<PathBuf>,
}

impl FakeAgent {
    pub fn new(events: Vec<AgentEvent>) -> Self {
        Self { events, file_writes: Vec::new(), workspace: None }
    }

    pub fn success() -> Self {
        Self::new(vec![
            AgentEvent::text("fake_1", "Task completed successfully", true),
            AgentEvent::turn_ended(TurnOutcome::Completed),
        ])
    }

    pub fn with_tool_call(tool_name: impl Into<String>, result: impl Into<String>) -> Self {
        let tool_name = tool_name.into();
        Self::new(vec![
            AgentEvent::Tool(ToolEvent::Result {
                result: ToolCallResult {
                    id: "fake_call_1".to_string(),
                    name: tool_name,
                    arguments: "{}".to_string(),
                    result: result.into(),
                },
                result_meta: None,
            }),
            AgentEvent::text("fake_2", "Task completed using tools", true),
            AgentEvent::turn_ended(TurnOutcome::Completed),
        ])
    }

    pub fn writes_file(path: impl Into<PathBuf>, contents: impl Into<String>) -> Self {
        Self::success().with_file_write(path, contents)
    }

    pub fn with_file_write(mut self, path: impl Into<PathBuf>, contents: impl Into<String>) -> Self {
        self.file_writes.push((path.into(), contents.into()));
        self
    }

    pub fn with_workspace(mut self, workspace: impl AsRef<Path>) -> Self {
        self.workspace = Some(workspace.as_ref().to_path_buf());
        self
    }
}

impl Agent for FakeAgent {
    fn run(&self, _task: Task) -> impl Stream<Item = AgentRunResult> + Send + '_ {
        try_stream! {
            if !self.file_writes.is_empty() && self.workspace.is_none() {
                Err(RunError::ConfigurationError("FakeAgent file writes require a bound workspace".to_string()))?;
            }

            for (path, contents) in &self.file_writes {
                let file_path = self.workspace.as_ref().expect("workspace checked above").join(path);
                if let Some(parent) = file_path.parent() {
                    create_dir_all(parent).await.map_err(|error| {
                        RunError::ExecutionFailed(format!("failed to create fake file parent: {error}"))
                    })?;
                }
                write(&file_path, contents)
                    .await
                    .map_err(|error| RunError::ExecutionFailed(format!("failed to write fake file: {error}")))?;
            }

            for event in &self.events {
                yield event.clone();
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use aether_core::events::{MessageEvent, TurnEvent};

    #[tokio::test]
    async fn fake_agent_success_sends_done() {
        let agent = FakeAgent::success();
        let transcript = crate::Transcript::from_stream(agent.run(Task::new("test task"))).await.unwrap();
        let messages = transcript.events();

        assert!(matches!(messages.first(), Some(AgentEvent::Message(MessageEvent::Text { .. }))));
        assert!(matches!(messages.get(1), Some(AgentEvent::Turn(TurnEvent::Ended { .. }))));
    }

    #[tokio::test]
    async fn fake_agent_with_tool_call_sends_tool_messages() {
        let agent = FakeAgent::with_tool_call("bash", "success");
        let transcript = crate::Transcript::from_stream(agent.run(Task::new("test task"))).await.unwrap();
        let messages = transcript.events();

        assert_eq!(messages.len(), 3);
        assert!(matches!(messages.last(), Some(AgentEvent::Turn(TurnEvent::Ended { .. }))));
    }

    #[tokio::test]
    async fn fake_agent_writes_file_in_workspace() {
        let temp_dir = tempfile::tempdir().unwrap();
        let agent = FakeAgent::writes_file("nested/hello.txt", "hello").with_workspace(temp_dir.path());
        crate::Transcript::from_stream(agent.run(Task::new("test task"))).await.unwrap();
        assert_eq!(std::fs::read_to_string(temp_dir.path().join("nested/hello.txt")).unwrap(), "hello");
    }
}

File: packages/aether-evals/src/FakeAgent.ts
import { mkdir, writeFile } from "node:fs/promises";
import { dirname, join } from "node:path";

import type { AgentEvent } from "@aether-agent/sdk";
import { AetherEvalError } from "./errors.js";
import type { Agent } from "./Agent.js";
import type { Task } from "./task.js";
import { turnEnded } from "./transcript.js";
import type { Workspace } from "./workspace.js";

/**
 * In-memory {@link Agent} for testing eval harnesses without Docker. Streams a scripted transcript
 * and optionally writes files into the workspace, mirroring how a real agent is observed.
 */
export class FakeAgent implements Agent {
  constructor(
    readonly events: AgentEvent[],
    readonly fileWrites: ReadonlyArray<readonly [string, string]> = [],
    readonly workspace: Workspace | undefined = undefined,
  ) {}

  static success(): FakeAgent {
    return new FakeAgent([
      {
        category: "message",
        event: {
          type: "text",
          message_id: "fake_1",
          chunk: "Task completed successfully",
          is_complete: true,
        },
      },
      turnEnded(),
    ]);
  }

  static withToolCall(toolName: string, result: string): FakeAgent {
    return new FakeAgent([
      {
        category: "tool",
        event: {
          type: "result",
          result: {
            id: "fake_call_1",
            name: toolName,
            arguments: "{}",
            result,
          },
        },
      },
      {
        category: "message",
        event: {
          type: "text",
          message_id: "fake_2",
          chunk: "Task completed using tools",
          is_complete: true,
        },
      },
      turnEnded(),
    ]);
  }

  static writesFile(path: string, contents: string): FakeAgent {
    return FakeAgent.success().withFileWrite(path, contents);
  }

  withFileWrite(path: string, contents: string): FakeAgent {
    return new FakeAgent(
      this.events,
      [...this.fileWrites, [path, contents]],
      this.workspace,
    );
  }

  withWorkspace(workspace: Workspace): FakeAgent {
    return new FakeAgent(this.events, this.fileWrites, workspace);
  }

  async *run(_task: Task): AsyncIterable<AgentEvent> {
    if (this.fileWrites.length > 0 && !this.workspace) {
      throw new AetherEvalError(
        "configuration_error",
        "FakeAgent file writes require a bound workspace",
      );
    }

    for (const [relativePath, contents] of this.fileWrites) {
      const target = join(this.workspace!.path, relativePath);
      await mkdir(dirname(target), { recursive: true });
      await writeFile(target, contents);
    }

    for (const event of this.events) {
      yield event;
    }
  }
}
